src/utils/leb128.ts

```diff
@@ -0,0 +1,80 @@
+/**
+ * Copyright (c) Hathor Labs and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+export interface Leb128DecodeResult {
+  value: bigint;
+  rest: Buffer;
+}
+
+export function encodeSigned(value: bigint | number, maxBytes: number | null = null): Buffer {
+  let val = BigInt(value);
+  const result: bigint[] = [];
+  while (true) {
+    const byte = val & 0b0111_1111n;
+    val >>= 7n;
+    if (
+      (val === 0n && (byte & 0b0100_0000n) === 0n) ||
+      (val === -1n && (byte & 0b0100_0000n) !== 0n)
+    ) {
+      result.push(byte);
+      if (maxBytes !== null && result.length > maxBytes) {
+        throw new Error(`Cannot encode more than ${maxBytes} bytes`);
+      }
+      // Need to convert the bigint values of `result` into Number to use with Buffer.
+      // This is a safe operation since the value of the elements can only go up to 0xFF
+      return Buffer.from(result.map(b => Number(b)));
```

Comment:
Why wouldn't this be a problem? Can't we have problems converting bigint values into Number?

Reply:
Yes, bigint values higher than [`Number.MAX_SAFE_INTEGER`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_SAFE_INTEGER) cannot be converted safely.
Since `MAX_SAFE_INTEGER` is `2^53 -1` it can happen with 8 byte integers (like our amount) but we can safely convert if the value cannot go up that high.